Add UpstreamPackages to docs.ros.org documentation (#3824)
ros2/ros2_documentation: source/The-ROS2-Project/Contributing/Developer-Guide.rst

Each package should be in a subfolder which has the same name as the package.
If a repository contains only a single package it can optionally be in the root of the repository.

Upstream Packages
^^^^^^^^^^^^^^^^^

Packages in Debian and Ubuntu Upstream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Thanks to diligent effort from Jochen Sprickerhof and Leopold Palomo-Avellaneda, some of the ROS 2 packages are now available from the main Debian and Ubuntu repositories.
`Here is a short overview of the process from Jochen at ROSCon 2015 <https://vimeo.com/142151399#t=29m15s>`_.
The original ROS packages have been modified to follow Debian guidelines, which includes splitting packages into multiple pieces, changing names in some cases, installing to /usr according to FHS guidelines, and using soversions on shared libraries.

> In addition several of the bootstrap dependencies such as command line tools like `vcstool` and `colcon` as well as some libraries like `osrf-pycommon` and `ament` are also packaged upstream.

Reviewer: ```suggestion
In addition several of the bootstrap dependencies such as command line tools like ``vcstool`` and ``colcon`` as well as some libraries like ``osrf-pycommon`` and ``ament`` are also packaged upstream.
```

> Unlike the OSRF-provided ROS packages from packages.ros.org, the packages in the upstream repositories are not attached to a specific :doc:`ROS distribution <../../Releases>`.

Reviewer: ```suggestion
Unlike the OSRF-provided ROS packages from http://packages.ros.org, the packages in the upstream repositories are not attached to a specific :doc:`ROS distribution <../../Releases>`.
```

Rather, they represent a snapshot in time  that will be updated periodically within Debian unstable and then latched at various points into downstream Debian and Ubuntu distributions.

Don't mix the streams
~~~~~~~~~~~~~~~~~~~~~

> We strongly recommend against mixing ROS packages from upstream Debian/Ubuntu and from packages.ros.org on the same system.

Reviewer: ```suggestion
We strongly recommend against mixing ROS packages from upstream Debian/Ubuntu and from http://packages.ros.org on the same system.
```

In some cases such a mixed system will work correctly, but there can be negative interactions between the two sets of packages.
We’re working with Jochen and friends to minimize the chance of problems via documentation and package conflict specifications, but we expect some risks to remain, including some fairly subtle issues.

> As such, we recommend that you choose to either install packages from upstream or from packages.ros.org, but not both.

Reviewer: ```suggestion
As such, we recommend that you choose to either install packages from upstream or from http://packages.ros.org, but not both.
```

> Not only should you not install packages from both at the same time, but if you intend to use the upstream packages (not from packages.ros.org) then you should not even have the packages.ros.org entries in your apt sources (i.e. in any files in ``/etc/apt/sources*``).

Reviewer: ```suggestion
Not only should you not install packages from both at the same time, but if you intend to use the upstream packages then you should not even have the http://packages.ros.org entries in your apt sources (i.e. in any files in ``/etc/apt/sources*``).
```

Each package should be in a subfolder which has the same name as the package.
If a repository contains only a single package it can optionally be in the root of the repository.

Upstream Packages
^^^^^^^^^^^^^^^^^

Packages in Debian and Ubuntu Upstream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Thanks to diligent effort from Jochen Sprickerhof and Leopold Palomo-Avellaneda, there are now ROS packages available upstream in Debian unstable and Ubuntu Xenial that can be installed from the main Debian and Ubuntu repositories.
`Here is a short overview from Jochen at ROSCon 2015 <https://vimeo.com/142151399#t=29m15s>`_.

> Most of the packages up to desktop-full are available (`overview of package status at Debian Science <https://wiki.debian.org/DebianScience/Robotics/ROS/Packages>`_). The original ROS packages have been modified to follow Debian guidelines, which includes splitting packages into multiple pieces, changing names in some cases, installing to /usr according to FHS guidelines, and using soversions on shared libraries. For example, roslib is called python-roslib and roscpp is in the libroscpp-dev and libroscpp0d packages.

Reviewer 2: maybe a little adjustment for ROS 2 from ROS 1?

```suggestion
Most of the packages up to desktop-full are available (`overview of package status at Debian Science <https://wiki.debian.org/DebianScience/Robotics/ROS/Packages>`_). The original ROS packages have been modified to follow Debian guidelines, which includes splitting packages into multiple pieces, changing names in some cases, installing to /usr according to FHS guidelines, and using soversions on shared libraries. For example, rclcpp is in the librclcpp-dev and librclcpp0d packages.
```

Author: rclcpp hasn't been ported yet. But I called this out to say ROS 1 specifically for ROS packages. But I believe that they're on the way over time.

Author: They have started the port for ROS 2. I added a link in [2e62c95](https://github.com/ros2/ros2_documentation/pull/3824/commits/2e62c9549e65fc3b686f462fc8d8f7bf42ca10ea)

> Unlike the OSRF-provided ROS packages from packages.ros.org, the packages in the upstream repositories are not attached to a specific `ROS distribution <https://docs.ros.org/en/rolling/Releases.html>`_.

Reviewer 2: ```suggestion
Unlike the OSRF-provided ROS packages from packages.ros.org, the packages in the upstream repositories are not attached to a specific `ROS distribution <https://docs.ros.org/en/{DISTRO}/Releases.html>`_.
```

Author: looks good. Manually applied as it didn't apply correctly when I'm trying out codespaces editing with other edits.

> Rather, they represent a rolling release that will be updated periodically within Debian unstable and then periodically latched at various points into downstream Debian and Ubuntu distributions.

Reviewer 2: ```suggestion
Rather, they represent a {DISTRO} release that will be updated periodically within Debian unstable and then periodically latched at various points into downstream Debian and Ubuntu distributions.
```

Author: THis isn't the proper name rolling but the genric.

Don't mix the streams
~~~~~~~~~~~~~~~~~~~~~
We strongly recommend against mixing ROS packages from upstream Debian/Ubuntu and from packages.ros.org on the same system.
In some cases such a mixed system will work correctly, but there can be negative interactions between the two sets of packages.

> We’re working with Jochen and friends to minimize the chance of problems via documentation and package conflict specifications, but we expect some risks to remain, including some fairly subtle issues.

Reviewer 2: Is this still work in progress? if not, we can delete this line?

Author: It's mostly stable, but there's always going to be a bit of a push and pull as things slowly evolve.

As such, we recommend that you choose to either install packages from upstream or from packages.ros.org, but not both.
Not only should you not install packages from both at the same time, but if you intend to use the upstream packages (not from packages.ros.org) then you should not even have the packages.ros.org entries in your apt sources (i.e. in any files in /etc/apt/sources*) as this can cause mixing of packages which overlap by name between the two sources, e.g. python-rospkg.

Known Differences
~~~~~~~~~~~~~~~~~
As compared to the ROS packages from packages.ros.org, there are some differences in the upstream ROS packages that people should be aware of:

* The package set is incomplete.
  Unsurprisingly, not all packages are yet available upstream. Currently the packages of desktop-full are available and we anticipate that more will become available over time.
* Packages may have different names and be partitioned differently.

> 

Reviewer 2: those lines are not necessary.

Each package should be in a subfolder which has the same name as the package.
If a repository contains only a single package it can optionally be in the root of the repository.

Upstream Packages
^^^^^^^^^^^^^^^^^

Packages in Debian and Ubuntu Upstream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Thanks to diligent effort from Jochen Sprickerhof and Leopold Palomo-Avellaneda, there are now ROS packages available upstream in Debian unstable and Ubuntu Xenial that can be installed from the main Debian and Ubuntu repositories.
> `Here is a short overview from Jochen at ROSCon 2015 <https://vimeo.com/142151399#t=29m15s>`_.
> 
> Most of the ROS 1 packages up to desktop-full are available (`overview of package status at Debian Science <https://wiki.debian.org/DebianScience/Robotics/ROS/Packages>`_).
> The original ROS packages have been modified to follow Debian guidelines, which includes splitting packages into multiple pieces, changing names in some cases, installing to /usr according to FHS guidelines, and using soversions on shared libraries.
> For example, roslib is called python-roslib and roscpp is in the libroscpp-dev and libroscpp0d packages.
> In addition several of the bootstrap dependencies such as command line tools like `vcstool` and `colcon` as well as some libraries like `osrf-pycommon` and `ament` are also packaged upstream.

Reviewer: ```suggestion
Thanks to diligent effort from Jochen Sprickerhof and Leopold Palomo-Avellaneda, some of the ROS 2 packages are now available from the main Debian and Ubuntu repositories.
`Here is a short overview of the process from Jochen at ROSCon 2015 <https://vimeo.com/142151399#t=29m15s>`_.
The original ROS packages have been modified to follow Debian guidelines, which includes splitting packages into multiple pieces, changing names in some cases, installing to /usr according to FHS guidelines, and using soversions on shared libraries.

In addition several of the bootstrap dependencies such as command line tools like `vcstool` and `colcon` as well as some libraries like `osrf-pycommon` and `ament` are also packaged upstream.
```

> Unlike the OSRF-provided ROS packages from packages.ros.org, the packages in the upstream repositories are not attached to a specific `ROS distribution <https://docs.ros.org/en/{DISTRO}/Releases.html>`_.

Reviewer: ```suggestion
Unlike the OSRF-provided ROS packages from packages.ros.org, the packages in the upstream repositories are not attached to a specific :doc:`ROS distribution <../../Releases>`_.
```

> Rather, they represent a rolling release that will be updated periodically within Debian unstable and then periodically latched at various points into downstream Debian and Ubuntu distributions.

Reviewer: ```suggestion
Rather, they represent a snapshot in time  that will be updated periodically within Debian unstable and then latched at various points into downstream Debian and Ubuntu distributions.
```

Don't mix the streams
~~~~~~~~~~~~~~~~~~~~~
We strongly recommend against mixing ROS packages from upstream Debian/Ubuntu and from packages.ros.org on the same system.
In some cases such a mixed system will work correctly, but there can be negative interactions between the two sets of packages.
We’re working with Jochen and friends to minimize the chance of problems via documentation and package conflict specifications, but we expect some risks to remain, including some fairly subtle issues.

As such, we recommend that you choose to either install packages from upstream or from packages.ros.org, but not both.

> Not only should you not install packages from both at the same time, but if you intend to use the upstream packages (not from packages.ros.org) then you should not even have the packages.ros.org entries in your apt sources (i.e. in any files in /etc/apt/sources*) as this can cause mixing of packages which overlap by name between the two sources, e.g. python-rospkg.

Reviewer: ```suggestion
Not only should you not install packages from both at the same time, but if you intend to use the upstream packages (not from packages.ros.org) then you should not even have the packages.ros.org entries in your apt sources (i.e. in any files in ``/etc/apt/sources*``).
Having both of them enabled can cause mixing of packages which overlap by name between the two sources, e.g. ``python3-rospkg``.
```

Known Differences
~~~~~~~~~~~~~~~~~
As compared to the ROS packages from packages.ros.org, there are some differences in the upstream ROS packages that people should be aware of:

* The package set is incomplete.

>   Unsurprisingly, not all packages are yet available upstream. Currently the packages of desktop-full are available and we anticipate that more will become available over time.

Reviewer: ```suggestion
```
